### PR TITLE
Deta Shell Plugin

### PR DESCRIPTION
**Pull request by smyja**
## Type of change
 

- [x] Created a new plugin

## Related Issue(s)
  

* Resolves: #
* Relates: #

### CODE REVIEW

**Comment by smyja** [comment]
> Thanks for your submission, @smyja. In order to move forward, could you share a few things with us?
> 
> - What is Deta about and what the CLI does? We'd really appreciate it if it could be added to the overview section of the PR description. (And ideally we'd like the "how to test" and "changelog" sections filled out as well)
> - The links `deta.com` don't seem to be correct. They redirect to a malicious site. Did you mean `deta.space` instead based on the [CLI repository](https://github.com/deta/space-cli)? 
> - Lastly, were you able to test this CLI somehow? Based on the CLI repository, Deta CLI's environment variable is `SPACE_ROOT` but the one proposed in this PR is `DETA_API_KEY`. I am not able to locate the latter in the CLI codebase so far. 
> 
> Ideally, we'd like that contributors build the shell plugin locally and test it before submitting a PR. If you are running into any issues with the setup or local testing, please share specifics of what's happening and we are happy to guide you!

This is a draft pr. I'm not done. It's deta.space. I'll set it to ready to review when I am done. Deta is a web hosting platform like vercel.

**Comment by arunsathiya** [comment]
Thank you for clarifying! We'll keep an eye out for your updates.

**Comment by techcraver** [comment]
Hello! Is this a submission for the 1Password Hackathon with Hashnode? If so, please be sure you write a blog post to make your submission official. Full [instructions](https://hashnode.com/hackathons/1password#participate) are on the Hackathon page.

**Comment by smyja** [comment]
> Hello! Is this a submission for the 1Password Hackathon with Hashnode? If so, please be sure you write a blog post to make your submission official. Full [instructions](https://hashnode.com/hackathons/1password#participate) are on the Hackathon page.

This is for Hashnode.

**Comment by techcraver** [comment]
Cool! We can't wait to see the blog post.

File: plugins/deta/api_key.go [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+		},
+		DefaultProvisioner: provision.EnvVars(defaultEnvVarMapping),
+		Importer: importer.TryAll(
+			importer.TryEnvVarPair(defaultEnvVarMapping),
+			TryDetaConfigFile(),
+		)}
+}
+
+var defaultEnvVarMapping = map[string]sdk.FieldName{
+	"DETA_API_KEY": fieldname.APIKey, // TODO: Check if this is correct
```
Comment (AndyTitu) [inline plugins/deta/api_key.go:43]: Could you be pointing me please to any documentation about the `DETA_API_KEY` env var?

I found that `space` cli uses a `SPACE_ACCESS_TOKEN` instead: https://github.com/deta/space-cli#customizing-the-api-endpoint

Comment (AndyTitu) [reply plugins/deta/api_key.go:43]: I also found a `DETA_PROJECT_KEY` but this seems to be for something else.

Finally, I found something about API Keys but still no reference to that env var: https://deta.space/docs/en/build/fundamentals/the-space-runtime/authentication#api-keys

**Review by AndyTitu — COMMENTED** [review]
@smyja  Very interesting tool! I have left a question for now to familiarise myself with the direction of this PR

Also, let's make sure all your commits are signed from the beginning as unsigned commits won't allow us to merge this PR

File: plugins/deta/api_key.go (hunk 1 above)
Comment (hculea) [reply plugins/deta/api_key.go:43]: I think `DETA_API_KEY` may be the name that the template generated. So definitely worth having another look over this provisioner.

Comment (smyja) [reply plugins/deta/api_key.go]: > I think `DETA_API_KEY` may be the name that the `make new-plugin` command generated. So definitely worth having another look over this provisioner.

yes it is.

**Comment by smyja** [comment]
> @smyja Very interesting tool! I have left a question for now to familiarise myself with the direction of this PR
> 
> Also, let's make sure all your commits are signed from the beginning as unsigned commits won't allow us to merge this PR

I have signed the commit and also removed the api key environment variable.

**Comment by smyja** [comment]
> Thanks for your submission, @smyja. In order to move forward, could you share a few things with us?
> 
>     * What is Deta about and what the CLI does? We'd really appreciate it if it could be added to the overview section of the PR description. (And ideally we'd like the "how to test" and "changelog" sections filled out as well)
> 
>     * The links `deta.com` don't seem to be correct. They redirect to a malicious site. Did you mean `deta.space` instead based on the [CLI repository](https://github.com/deta/space-cli)?
> 
>     * Lastly, were you able to test this CLI somehow? Based on the CLI repository, Deta CLI's environment variable is `SPACE_ROOT` but the one proposed in this PR is `DETA_API_KEY`. I am not able to locate the latter in the CLI codebase so far.
> 
> 
> Ideally, we'd like that contributors build the shell plugin locally and test it before submitting a PR. If you are running into any issues with the setup or local testing, please share specifics of what's happening and we are happy to guide you!

 the space tokens file doesn't have a file extension in it, `detaspace/space_tokens` but it contains a dictionary. Should i assume it is json?

**Comment by arunsathiya** [comment]
@smyja Are you perhaps referring to a [INI file](https://en.wikipedia.org/wiki/INI_file)? There's also a `ToINI` parser that you can use:

https://github.com/1Password/shell-plugins/blob/a39d64acfb1dd3cdd496507d54cd4c8306da6f78/sdk/importer/file_importer.go#L81-L88

**Comment by smyja** [comment]
> @smyja Are you perhaps referring to a [INI file](https://en.wikipedia.org/wiki/INI_file)? There's also a `ToINI` parser that you can use:
> 
> https://github.com/1Password/shell-plugins/blob/a39d64acfb1dd3cdd496507d54cd4c8306da6f78/sdk/importer/file_importer.go#L81-L88

It's not a .ini file. There's actually no extension on the file. just spacetokens. When I use nano, I can see the API key in it.

**Comment by hculea** [comment]
If it only contains the APIKey, you can load the entire file into a field:
```
func TryDetaConfigFile(path string) sdk.Importer {
	return importer.TryFile(path, func(ctx context.Context, contents importer.FileContents, in sdk.ImportInput, out *sdk.ImportAttempt) {
		credential := contents.ToString()
		fields := make(map[sdk.FieldName]string)
		if len(credential) != 0 {
			fields[fieldname.APIKey] = credential
		}
		out.AddCandidate(sdk.ImportCandidate{
			Fields: fields,
		})
	})
}
```

**Comment by hculea** [comment]
Hey @smyja , I see there hasn't been any activity on this PR in a bit, is there anything that we can help you with?

We'd love to get this to the finish line together! 🚀

**Comment by smyja** [comment]
> Hey @smyja , I see there hasn't been any activity on this PR in a bit, is there anything that we can help you with?
> 
> We'd love to get this to the finish line together! 🚀

Hey, i won't be proceeding with this due to the number of entries for the category.

**Pull request closed without merging**